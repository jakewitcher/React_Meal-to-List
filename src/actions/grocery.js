export const onAddGroceryAsync = (
    {
        name = '',
        items = [],
    } = {}
) => ({
    type: 'ADD_GROCERY_ASYNC',
    grocery: {
        name,
        items,
    }
});

export const onDeleteGroceryAsync = (id = '') => ({
    type: 'DELETE_GROCERY_ASYNC',
    id,
});

export const onEditGroceryAsync = (
    {
        name = '',
        id = '',
        items = [],
    } = {}
) => ({
    type: 'EDIT_GROCERY_ASYNC',
    grocery: {
        id,
        name,
        items,
    }
});

export const onSetGroceryAsync = () => ({
    type: 'SET_GROCERY_ASYNC',
});