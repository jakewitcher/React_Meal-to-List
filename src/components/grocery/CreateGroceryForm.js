import React from 'react';

const CreateGroceryForm = (props) => {
    const {
        selectedMealChange,
        selectedMeal,
        meals,
        groceryName,
        itemsList,
        groceryNameChange,
        addMealToList,
        resetGrocery,
        dispatch,
    } = props.groceryProps;
    const { addGrocery } = props;
    return (
        <div className="form">
            <h2 className="form__header">Create a New Grocery List</h2>
            <div className="form__form">
                <div className="form__nameinputfield">
                    <p>Grocery List Name</p>
                    <input
                        className="form__nameinputfield--input"
                        type="text"
                        value={groceryName}
                        onChange={groceryNameChange}
                    />
                </div>
                <div className="form__iteminputfields">
                    <div>
                        <p>Meals</p>
                        <select
                            className="form__iteminputfields--input form__iteminputfields--dropdown"
                            name="selectedMeal"
                            value={selectedMeal.name}
                            onChange={selectedMealChange}
                        >
                            {
                                meals.mealList.map(meal => <option value={meal.name} key={meal.id}>{meal.name}</option>)
                            }
                        </select>
                    </div>
                    <div className="form__button">
                        <button
                            className="form__button--add"
                            onClick={addMealToList}
                        >
                            Add Meal
                    </button>
                    </div>
                </div>
                <div>
                    <button
                        className="form__button--save"
                        onClick={() => {
                            dispatch(addGrocery({
                                name: groceryName,
                                items: itemsList,
                            }));
                            resetGrocery();
                        }
                        }
                    >
                        Save Grocery List
                </button>
                </div>
            </div>

        </div>
    )
}

export default CreateGroceryForm;