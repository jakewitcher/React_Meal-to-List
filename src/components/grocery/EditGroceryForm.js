import React from 'react';

const EditGroceryForm = (props) => {
    const {
        selectedMealChange,
        selectedMeal,
        meals,
        groceryName,
        itemsList,
        groceryNameChange,
        addMealToList,
        resetGrocery,
        dispatch,
    } = props.groceryProps;
    const { editGrocery, match } = props;
    return (
        <div className="form">
            <h2 className="form__header">Edit Grocery List</h2>
            <div className="form__form">
                <div className="form__nameinputfield">
                    <p>Grocery List Name</p>
                    <input
                        className="form__nameinputfield--input"
                        type="text"
                        value={groceryName}
                        onChange={groceryNameChange}
                    />
                </div>
                <div className="form__iteminputfields">
                    <div>
                        <p>Meals</p>
                        <select
                            className="form__iteminputfields--input form__iteminputfields--dropdown"
                            name="selectedMeal"
                            value={selectedMeal.name}
                            onChange={selectedMealChange}
                        >
                            {
                                meals.mealList.map(meal => <option value={meal.name} key={meal.id}>{meal.name}</option>)
                            }
                        </select>
                    </div>
                    <div className="form__button">
                        <button
                            className="form__button--add"
                            onClick={addMealToList}
                        >
                            Add Meal
                    </button>
                    </div>
                </div>
                <div>
                    <button
                        className="form__button--save"
                        onClick={() => {
                            const id = match.params.groceryId;
                            dispatch(editGrocery({
                                name: groceryName,
                                items: itemsList,
                                id,
                            }));
                            resetGrocery();
                        }
                        }
                    >
                        Save Grocery List
                </button>
                </div>
            </div>

        </div>
    )
}

export default EditGroceryForm;