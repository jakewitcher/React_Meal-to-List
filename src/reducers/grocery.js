const groceryDefaultState = {
    groceryList: [],
};

export default (state = groceryDefaultState, action) => {
    switch (action.type) {
        case 'ADD_GROCERY':
            return {
                ...state,
                groceryList: [...state.groceryList, action.grocery]
            };
        default:
            return state;
    }
}