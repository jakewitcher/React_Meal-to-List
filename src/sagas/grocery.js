import { put, takeEvery, select } from "redux-saga/effects";
import database from "../firebase/firebase";

const getUserId = state => state.auth.uid;

const formatItemList = list => {
  return list.reduce((list, item) => {
    list[item.itemName] = {
      amount: item.amount,
      unit: item.unit
    };
    return list;
  }, {});
};

function* addGrocery({ grocery = {} }) {
  const { itemList, name, groceryListMeals } = grocery;
  const items = formatItemList(itemList);
  const meals = groceryListMeals;

  let id;
  const uid = yield select(getUserId);

  yield database
    .ref(`users/${uid}/groceryLists`)
    .push({ name, items, meals })
    .then(ref => {
      id = ref.key;
    });

  yield put({
    type: "ADD_GROCERY",
    grocery: {
      name,
      itemList,
      groceryListMeals,
      id
    }
  });
}

function* editGrocery({ grocery = {} }) {
  const { itemList, name, id, groceryListMeals } = grocery;
  const items = formatItemList(itemList);
  const uid = yield select(getUserId);
  const meals = groceryListMeals;

  yield database
    .ref(`users/${uid}/groceryLists/${id}`)
    .update({ name, items, meals });

  yield put({
    type: "EDIT_GROCERY",
    grocery: {
      name,
      itemList,
      groceryListMeals,
      id
    }
  });
}

function* deleteGrocery({ id = "" }) {
  const uid = yield select(getUserId);

  yield database.ref(`users/${uid}/groceryLists/${id}`).remove();

  yield put({
    type: "DELETE_GROCERY",
    id
  });
}

function* setGrocery() {
  const lists = [];
  const uid = yield select(getUserId);

  yield database
    .ref(`users/${uid}/groceryLists`)
    .once("value")
    .then(snapshot => {
      snapshot.forEach(childSnapshot => {
        const itemList = [];
        childSnapshot.child("items").forEach(item => {
          itemList.push({
            itemName: item.key,
            amount: item.val().amount,
            unit: item.val().unit
          });
        });

        lists.push({
          id: childSnapshot.key,
          name: childSnapshot.val().name,
          itemList,
          groceryListMeals: childSnapshot.val().meals
        });
      });
    });
  yield put({ type: "SET_GROCERY", lists });
}

export function* watchAddGrocery() {
  yield takeEvery("ON_ADD_GROCERY", addGrocery);
}

export function* watchEditGrocery() {
  yield takeEvery("ON_EDIT_GROCERY", editGrocery);
}

export function* watchDeleteGrocery() {
  yield takeEvery("ON_DELETE_GROCERY", deleteGrocery);
}

export function* watchSetGrocery() {
  yield takeEvery("ON_SET_GROCERY", setGrocery);
}
