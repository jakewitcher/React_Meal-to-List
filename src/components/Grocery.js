import React from "react";
import { connect } from "react-redux";
import { Link } from "react-router-dom";
import { onDeleteGrocery } from "../store/grocery/actions";

export const Grocery = props => {
  const { name, id, itemList } = props;
  return (
    <div className="list-type">
      <Link className="list-type__link" to={`grocerylists/grocery-list/${id}`}>
        <p className="list-type__name">{name}</p>
      </Link>
      <p className="list-type__details">
        number of items:{" "}
        <span className="list-type__details list-type__details--span">
          {itemList.length}
        </span>
      </p>
      <div className="list-type__update">
        <Link className="list-type__link" to={`grocerylists/edit/${id}`}>
          <i className="list-type__edit fas fa-edit" />
        </Link>
        <i
          className="list-type__delete fas fa-trash"
          onClick={() => props.onDeleteGrocery(id)}
        />
      </div>
    </div>
  );
};

const mapDispatchToProps = dispatch => {
  return {
    onDeleteGrocery: id => dispatch(onDeleteGrocery(id))
  };
};

export default connect(
  undefined,
  mapDispatchToProps
)(Grocery);
