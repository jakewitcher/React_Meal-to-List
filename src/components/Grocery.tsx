import React from "react";
import { Dispatch } from "redux";
import { connect } from "react-redux";
import { Link } from "react-router-dom";
import { onDeleteGrocery } from "../store/grocery/actions";
import { Item } from "../store/item/types";

interface IGroceryProps {
  name: string;
  id: string;
  itemList: Item[];
  onDeleteGrocery(id: string): void;
}

export const Grocery: React.FC<IGroceryProps> = props => {
  const { name, id, itemList } = props;
  return (
    <div className="list-type">
      <Link className="list-type__link" to={`grocerylists/grocery-list/${id}`}>
        <p className="list-type__name">{name}</p>
      </Link>
      <p className="list-type__details">
        number of items:{" "}
        <span className="list-type__details list-type__details--span">
          {itemList.length}
        </span>
      </p>
      <div className="list-type__update">
        <Link className="list-type__link" to={`grocerylists/edit/${id}`}>
          <i className="list-type__edit fas fa-edit" />
        </Link>
        <i
          className="list-type__delete fas fa-trash"
          onClick={() => props.onDeleteGrocery(id)}
        />
      </div>
    </div>
  );
};

const mapDispatchToProps = (dispatch: Dispatch) => {
  return {
    onDeleteGrocery: (id: string) => dispatch(onDeleteGrocery(id))
  };
};

export default connect(
  undefined,
  mapDispatchToProps
)(Grocery);
