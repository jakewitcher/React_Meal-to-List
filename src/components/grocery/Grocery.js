import React from 'react';
import { connect } from 'react-redux';
import { deleteGrocery } from '../../actions/grocery';

const Grocery = (props) => {
    const { name, dispatch, editExistingGrocery, id } = props;
    return (
        <div>
            <p>{name}</p>
            <i 
                className="fas fa-edit"
                onClick={() => editExistingGrocery()}
            />
            <i 
                className="fas fa-trash"
                onClick={() => dispatch(deleteGrocery(id))}    
            />
        </div>
    );
}

export default connect()(Grocery);