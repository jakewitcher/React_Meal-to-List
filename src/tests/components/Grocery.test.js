import React from "react";
import { shallow } from "enzyme";
import { Grocery } from "../../components/grocery/Grocery";
import { groceryLists } from "../fixtures/grocery";

let name, id, itemList, wrapper, onDeleteGrocery;

beforeEach(() => {
  name = groceryLists[0].name;
  id = groceryLists[0].id;
  itemList = groceryLists[0].items;
  onDeleteGrocery = jest.fn();
  wrapper = shallow(
    <Grocery
      name={name}
      id={id}
      itemList={itemList}
      onDeleteGrocery={onDeleteGrocery}
    />
  );
});

test("should render Grocery correctly", () => {
  expect(wrapper).toMatchSnapshot();
});

test("should delete grocery list by id", () => {
  wrapper
    .find("i")
    .at(1)
    .simulate("click");
  expect(onDeleteGrocery).toHaveBeenLastCalledWith(id);
});
