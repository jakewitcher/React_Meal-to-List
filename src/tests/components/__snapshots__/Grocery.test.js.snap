// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Grocery correctly 1`] = `
<div
  className="list-type"
>
  <Link
    className="list-type__link"
    replace={false}
    to="grocerylists/grocery-list/201"
  >
    <p
      className="list-type__name"
    >
      Monday Meals
    </p>
  </Link>
  <p
    className="list-type__details"
  >
    number of items:
     
    <span
      className="list-type__details list-type__details--span"
    >
      8
    </span>
  </p>
  <div
    className="list-type__update"
  >
    <Link
      className="list-type__link"
      replace={false}
      to="grocerylists/edit/201"
    >
      <i
        className="list-type__edit fas fa-edit"
      />
    </Link>
    <i
      className="list-type__delete fas fa-trash"
      onClick={[Function]}
    />
  </div>
</div>
`;
